### PR TITLE
Fix gh actions

### PR DESCRIPTION
**Pull request by mauwii**
in favor of #1118 🙈

### CODE REVIEW

**Comment by mauwii** [comment]
@lstein @tildebyte @Any-Winter-4079 
Here you go ;P

File: environment-mac.yml [hunk 1]
```diff
@@ -3,56 +3,54 @@ channels:
   - pytorch
   - conda-forge
 dependencies:
-  - python==3.9.13
-  - pip==22.2.2
+  - python>=3.10, <3.11
+  - pip>=22
```
Comment (tildebyte) [inline environment-mac.yml:7]: Whoops! Almost missed this (and it's even my fault - I didn't know better at the time) - 'pip' should never be in the dependencies.

**Review by tildebyte — CHANGES_REQUESTED** [review]
AFAICT, 'create-caches.yml' has a couple of issues:

- "Use Cached Stable Diffusion v1.4 Model" is duplicated
- "Use cached Conda Environment" and "Use cached Conda Packages" look to be almost exactly the same

Also, please unpin 'environment.yml'

File: environment-mac.yml (hunk 1 above)
Comment (mauwii) [reply environment-mac.yml:7]: I am not used to Conda, so I just did what felt right and changed it to use the most recent one 🙈

Comment (tildebyte) [reply environment-mac.yml:7]: Same! I only understood much later that it shouldn't be done :)

**Comment by mauwii** [comment]
> AFAICT, 'create-caches.yml' has a couple of issues:
> 
> * "Use Cached Stable Diffusion v1.4 Model" is duplicated
> * "Use cached Conda Environment" and "Use cached Conda Packages" look to be almost exactly the same
> 
> Also, please unpin 'environment.yml'

May I ask what's the purpose behind "create-caches.yml" at all? maybe we could even delete it completeley, going to unpin versions in environment.yml then, but since it had no problems (even works on Linux and Windows) and version pinning is a good thing in the end (when it is working for everybody ofc), I would have left them pinned until further solution is worked out.

**Comment by mauwii** [comment]
> AFAICT, 'create-caches.yml' has a couple of issues:
> 
> * "Use Cached Stable Diffusion v1.4 Model" is duplicated
> * "Use cached Conda Environment" and "Use cached Conda Packages" look to be almost exactly the same
> 
> Also, please unpin 'environment.yml'

Conda is actually right, it is caching Environment and Packages

- Use cached Conda Environment
- Use cached Conda Packages

Environment caching is not best practice, but since I thought it would run more than it will do now, it made totally sence (since conda pkgs/environment won't change often, but Code would).

**Comment by mauwii** [comment]
@tildebyte to come back to the "bare minimal test", imho the test which is running is pretty huge: 

all but refs/heads/main:

- `banana sushi -Ak_lms -S42`
this could be broken town to something with only 10 iterations when testing a PR

refs/heads/main:

- `banana sushi -Ak_lms -S42`
- `banana sushi -Addim -S42`
- `banana sushi -Ak_lms -W640 -H480 -S42`
- `banana sushi -Ak_lms -S42 -G1 -U 2 0.5`
- `banana sushi -Ak_lms -S42 -v0.2 -n3`
- `banana sushi -Ak_lms -S42 -V1349749425:0.1,4145759947:0.1`
- `snake -I outputs/preflight/000006.4145759947.png -S42`
- `snake -I outputs/preflight/000006.4145759947.png -S42 -W640 -H640 --fit`
- `strawberry sushi -I./image-and-mask.png -S42 -f0.9 -s100 -C15`

this does a lot (but is fine to make sure it is still working ofc, but I also agree that for validating a PR this is pure overkill)

This can be changed by editing the two files in tests/*prompts.txt, also can be advanced ofc

**Comment by mauwii** [comment]
What - how can I remove a requested review xD Sorry, did only pres the "Re-require review" button to point out that the asked changes are implemented 🙆🏻‍♂️

**Comment by mauwii** [comment]
this would be `banana sushi -Ak_lms -S42 -s10`:

image

imho this is far enough to do a quick PR validation while taking only 20% of the time 🤓

**Comment by mauwii** [comment]
> Also, please unpin 'environment.yml'

needed to revert this since unpinned environment yml failed:

```shell
Installing pip dependencies: ...working... ::warning::Pip subprocess error:%0AERROR: Could not detect requirement name for 'git+https://github.com/openai/CLIP.git@main#egg', please specify one with #egg=your_package_name%0A%0A
Pip subprocess error:
ERROR: Could not detect requirement name for 'git+https://github.com/openai/CLIP.git@main#egg', please specify one with #egg=your_package_name

Ran pip subprocess with arguments:
['/usr/share/miniconda3/envs/invokeai/bin/python', '-m', 'pip', 'install', '-U', '-r', '/home/runner/work/InvokeAI/InvokeAI/condaenv.6u8ex5u2.requirements.txt']
Pip subprocess output:

failed
Warning: 
CondaEnvException: Pip failed
```

**Comment by mauwii** [comment]
So atm we have:

- 35 minutes without cache (like f.E. a changed environment.yml)
- 20 minutes with cache

which imho should be totally worth the time vs merging something which then breaks buildability, reverting this, ....

@tildebyte @lstein @Any-Winter-4079 What's your opinion?

**Comment by mauwii** [comment]
Pip should not be completley removed from the environment.yml:

> Warning: you have pip-installed dependencies in your environment file, but you do not list pip itself as one of your conda dependencies.  Conda may not use the correct pip to install your packages, and they may end up in the wrong place.  Please add an explicit pip dependency.  I'm adding one for you, but still nagging you.

**Comment by tildebyte** [comment]
> May I ask what's the purpose behind "create-caches.yml" at all?

AFAIU (I didn't create it) it's to cache the model (possibly not needed as previously discussed), and packages which Conda would download (only useful if packages are pinned, so only *potentially* useful on 'main'.

**Comment by tildebyte** [comment]
> ```shell
> Installing pip dependencies: ...working... ::warning::Pip subprocess error:%0AERROR: Could not detect requirement name for 'git+https://github.com/openai/CLIP.git@main#egg', please specify one with #egg=your_package_name%0A%0A
> Pip subprocess error:
> ERROR: Could not detect requirement name for 'git+https://github.com/openai/CLIP.git@main#egg', please specify one with #egg=your_package_name
> 
> Ran pip subprocess with arguments:
> ['/usr/share/miniconda3/envs/invokeai/bin/python', '-m', 'pip', 'install', '-U', '-r', '/home/runner/work/InvokeAI/InvokeAI/condaenv.6u8ex5u2.requirements.txt']
> Pip subprocess output:
> 
> failed
> Warning: 
> CondaEnvException: Pip failed
> ```

Right, that's down to the method which pip uses to install from source repos. If the repo isn't installable (no setup.py at a minimum - there may be other factors), then pip **requires** an egg name - it's actually nothing to do with pinning.

...aaand in fact I just realized that this will completely break Conda installs; the following is REQUIRED

```python
      - -e git+https://github.com/CompVis/taming-transformers.git@master#egg=taming-transformers
      - -e git+https://github.com/openai/CLIP.git@main#egg=clip
      - -e git+https://github.com/Birch-san/k-diffusion.git@mps#egg=k_diffusion
      - -e git+https://github.com/TencentARC/GFPGAN.git#egg=gfpgan
      - -e .
```

**NOTE**: All of the rest of the desired unpinning for both Conda environment files should still be possible (although be aware that there might be some deps which require specific versions)!

**Comment by tildebyte** [comment]
> @tildebyte to come back to the "bare minimal test", imho the test which is running is pretty huge:

Oh, that's a really good point...

Sadly, that's not the only factor, e.g. [this run](https://github.com/invoke-ai/InvokeAI/actions/runs/3262192536/jobs/5358819118) took 33 minutes to **fail** (during Conda install) - the number of tests/number of steps would have made no difference here :(

**Comment by tildebyte** [comment]
@mauwii; I guess TL;DR add a single generation test for PRs, and we'll see how it goes.

I get that runtimes are free for open-source projects, but they still require electricity, and the planet is already starting to roast...

**Comment by mauwii** [comment]
> > @tildebyte to come back to the "bare minimal test", imho the test which is running is pretty huge:
> 
> Oh, that's a really good point...
> 
> Sadly, that's not the only factor, e.g. [this run](https://github.com/invoke-ai/InvokeAI/actions/runs/3262192536/jobs/5358819118) took 33 minutes to **fail** (during Conda install) - the number of tests/number of steps would have made no difference here :(

So it is better to first merge it and then realize in main branch "oops, it is not buildable anymore". Also it took you like 20h to answer, which could be like 40 runs, which in the end could f.E. help the developer to fix some bugs "last minute" which could otherwise have been missed and merged.

and of course the Number of tests only make a difference if conda does not fail. To verify the code it is running invoke.py with a prompt file, so if f.E. the preflight prompt is used, it would take hours to finish (since those free runners dont have much power and there are 9 prompts in it, even some bigger ones with resized images or 100 iterations). While running one test with only 10 prompts is taking about 5 minutes (but can you save lots of trouble).

And only point to a failed 30 minutes run, but leaving out the 20 minut runs which don't fail is also a bit odd....

So could you please explain why it is bad if the run fails after 30 minutes, which would prevent merging breaking changes into main branch. And just to mention it, as a administrator you could always configure the branch protection for admins to be skipable, so you wouldn't need to wait till it is finished if you want to live dangerous ;P

**Comment by mauwii** [comment]
> @mauwii; I guess TL;DR add a single generation test for PRs, and we'll see how it goes.
> 
> I get that runtimes are free for open-source projects, but they still require electricity, and the planet is already starting to roast...

ok, the last argument is a good one, but just to make something sure: I am reading comments from top to bottom, no TL:DR on my side .....

**Comment by tildebyte** [comment]
OK, I'm convinced. A single test generation[1] for PRs is a good idea.

[1] I mean "generated image" rather than e.g. "number of generation steps"

**Comment by mauwii** [comment]
should be ready to merge @tildebyte

**Review by tildebyte — APPROVED** [review]

**Comment by mauwii** [comment]
I hope I did the rebase correct now 🙈

**Pull request merged**